src/main.rs

```diff
@@ -51,43 +117,46 @@ fn main() {
     window.add(&entry);
     window.set_border_width(0);
     window.show_all();
-    let the_completions: Arc<Mutex<Box<Iterator<Item = String>>>> = Arc::new(Mutex::new(Box::new(vec![].into_iter())));
+    let the_completions: Arc<Mutex<Box<Iterator<Item = Completion>>>> = Arc::new(Mutex::new(Box::new(vec![].into_iter())));
+    let the_current_completion: Arc<Mutex<Box<Option<Completion>>>> = Arc::new(Mutex::new(Box::new(None)));
     window.connect_key_press_event(move |_, key| {
         let completions = the_completions.clone();
+        let current_completion = the_current_completion.clone();
         let keyval = key.keyval as i32;
         let keystate = (*key).state;
         debug!("key pressed: {}", keyval);
         match keyval {
             key::Escape => gtk::main_quit(),
             key::Return => {
-                let cmd = entry.get_text().unwrap();
                 debug!("keystate: {:?}", keystate);
                 debug!("Controlmask == {:?}", modifier_type::ControlMask);
+                let query = entry.get_text().unwrap();
+                let comp = *current_completion.lock().unwrap().clone();
+                let the_completion = comp.unwrap_or(engine.get_completions(&query).next().unwrap().to_owned());
+                let output = engine.run_completion(&the_completion).unwrap();
                 if keystate.intersects(modifier_type::ControlMask) {
                     debug!("ctrl pressed!");
-                    let output = execution::execute(cmd, false);
```

Comment:
Here i found a bug:

running something which should keep running (try chromium), it keeps rrun open..
Here the code checked if the used pressed "Ctrl-Return" and kept rrun open to display the return value of the command.
In case no Ctrl was pressed, the command must be spawned in the background (which is the forget_stdout parameter to execute())...
forget_stdout should probably be renamed to spawn_in_background or similar..